tests/h2-tests/tests/client_request.rs

```diff
@@ -84,431 +90,429 @@ fn recv_invalid_server_stream_id() {
     builder.max_concurrent_streams(1);
 
-    let h2 = builder
-        .handshake::<_, Bytes>(io)
-        .expect("handshake")
-        .and_then(|(mut client, h2)| {
-            let request = Request::builder()
-                .method(Method::GET)
-                .uri("https://example.com/")
-                .body(())
-                .unwrap();
-
-            let (response, _) = client.send_request(request, true).unwrap();
-            h2.drive(response).map(move |(h2, _)| (client, h2))
-        });
+    let h2 = async move {
+        let (mut client, mut h2) = builder.handshake::<_, Bytes>(io).await.unwrap();
+        let request = Request::builder()
+            .method(Method::GET)
+            .uri("https://example.com/")
+            .body(())
+            .unwrap();
+        let (response, _) = client.send_request(request, true).unwrap();
+        h2.drive(response).await.unwrap();
+    };
 
-    h2.join(srv).wait().expect("wait");
+    join(srv, h2).await;
 }
 
-#[test]
-fn request_over_max_concurrent_streams_errors() {
+#[tokio::test]
+async fn request_over_max_concurrent_streams_errors() {
     let _ = env_logger::try_init();
-    let (io, srv) = mock::new();
-
-
-    let srv = srv.assert_client_handshake_with_settings(frames::settings()
-                // super tiny server
-                .max_concurrent_streams(1))
-        .unwrap()
-        .recv_settings()
-        .recv_frame(
+    let (io, mut srv) = mock::new();
+
+    let srv = async move {
+        let settings = srv
+            .assert_client_handshake_with_settings(
+                frames::settings()
+                    // super tiny server
+                    .max_concurrent_streams(1),
+            )
+            .await;
+        assert_default_settings!(settings);
+        srv.recv_frame(
             frames::headers(1)
                 .request("POST", "https://example.com/")
                 .eos(),
         )
-        .send_frame(frames::headers(1).response(200).eos())
-        .recv_frame(frames::headers(3).request("POST", "https://example.com/"))
-        .send_frame(frames::headers(3).response(200))
-        .recv_frame(frames::data(3, "hello").eos())
-        .send_frame(frames::data(3, "").eos())
-        .recv_frame(frames::headers(5).request("POST", "https://example.com/"))
-        .send_frame(frames::headers(5).response(200))
-        .recv_frame(frames::data(5, "hello").eos())
-        .send_frame(frames::data(5, "").eos())
-        .close();
-
-    let h2 = client::handshake(io)
-        .expect("handshake")
-        .and_then(|(mut client, h2)| {
-            // we send a simple req here just to drive the connection so we can
-            // receive the server settings.
-            let request = Request::builder()
-                .method(Method::POST)
-                .uri("https://example.com/")
-                .body(())
-                .unwrap();
+        .await;
+        srv.send_frame(frames::headers(1).response(200).eos()).await;
```

Comment:
This isn't too big of a deal, but thought I'd bring it up: I'm partial to the chaining API that existed before, as it seemed a little more succinct. Is it difficult to keep?

cc @carllerche @hawkw @sgrif

Reply:
I think the tests reflect the future usage. IMO async-await is much clearer than chaining. In fact combinators based chaining API is main reason I abandoned a project I started 2 years back. Async-await has finally made it look like regular code. It is just my opinion though. 

Also, Tests were the major part of this work. The actual lib update was trivial.

Please let me know if you'd like me to revert it.

Reply:
It's alright, we can keep the new way. Thinking about in isolation, it does seem like one would expose these as async fns now. I just noticed what appeared to be noise adding all the `await`s.